### PR TITLE
Fix dyncast

### PR DESCRIPTION
**Pull request by UplinkCoder**
previously we could not really determine if we had a node of type statement.
also dyncast_object was never used.
now we can final switch on the dyncast value and make a few ast-matching operatios cheaper.

@WalterBright could you tell me what you think about this change.

### CODE REVIEW

**Comment by WalterBright** [comment]
It looks fine, I'll pull it after it passes.

**Comment by UplinkCoder** [comment]
@WalterBright it passed.
The cricle-ci script, is broken and its error cannot be relied upon.

**Comment by WalterBright** [comment]
Auto-merge toggled on

File: src/root/rootobject.d [hunk 1]
```diff
@@ -12,6 +12,8 @@ module ddmd.root.rootobject;
 
 import core.stdc.stdio;
 
+import ddmd.globals;
```
Comment (yebblies) [inline src/root/rootobject.d]: This is wrong, root should not be importing from dmd proper.  DYNCAST should be moved here instead.

File: src/expression.d [hunk 2]
```diff
@@ -2563,7 +2563,7 @@ extern (C++) abstract class Expression : RootObject
     }
 
     // kludge for template.isExpression()
-    override final int dyncast()
+    override final DYNCAST dyncast()
```
Comment (yebblies) [inline src/expression.d]: Why is this one not const?

Comment (UplinkCoder) [inline src/expression.d]: I overlooked it

File: src/root/rootobject.d (hunk 1 above)
Comment (UplinkCoder) [inline src/root/rootobject.d]: frankly I think rootobject should not be in root.

**Pull request merged**